ClusterResourceOverride: overcommit admission controller

sosiouxme opened the pull request:
Allows master config to override request and limit on submitted pods in order to control "request" overcommit and peg cpu limit to memory limit.

https://trello.com/c/rRN7m0zv/546-5-override-container-resource-request-based-on-limit

sosiouxme commented:
@csrwng I wasn't sure where to put this so I made a new directory. Can you think of a better location?

csrwng commented:
@sosiouxme yeah, this isn't for any of our resources. My preference would be under pkg/cmd/server/admission 
@deads2k do you have a preference?

deads2k commented on pkg/pod/admission/limit_request.go:
```diff
@@ -0,0 +1,98 @@
+}
+type podLimitRequest struct {
+	*admission.Handler
+	limitRequestConfig
```
Give this an explicit name.  It will make a future transition to config cleaner.

deads2k commented on pkg/pod/admission/limit_request.go:
```diff
@@ -0,0 +1,98 @@
+	}, err
+}
+
+func ReadConfig(config io.Reader) (limitRequestConfig, error) {
```
Follow @csrwng's examples for using versioned config objects.

deads2k commented on pkg/pod/admission/limit_request.go:
```diff
@@ -0,0 +1,98 @@
+	// adjust values as configured; no need to actually validate anything
+	for _, container := range pod.Spec.Containers {
+		resources := container.Resources
+		memLimit := resources.Limits.Memory()
```
scope into the `if` block

deads2k commented:
> @deads2k do you have a preference?

I would probably make a quota package: `pkg/quota/admission`.  At some point, we'll have additional quota admission plugins for handling quota of our resources.  It's a pretty minor change to move later though.

liggitt commented on pkg/pod/admission/limit_request.go:
```diff
@@ -0,0 +1,98 @@
+	if a.Enabled == false || !ok {
+		return nil // only interested in pods
+	}
+	// adjust values as configured; no need to actually validate anything
```
is this statement counting on this admission plugin coming after the LimitRanger one, so that whatever limit we set will not cause problems? if so, that might we worth calling out

sosiouxme replied:
Yes, it is supposed to come after, and yes that is worth a comment. Actually my reasoning was that in most cases the user isn't going to set any limit/request and I'm relying on the LimitRanger plugin to set values according to project limits object(s). I don't know if we're actually concerned about the adjustments landing outside what the limits say; I would think not.

liggitt replied:
well, if this comes before the limitranger admission, and it makes an adjustment that makes it fall outside the allowed range, then the pod would be rejected. it's just a bit odd to stack admission plugins, one of which could do something another one would find invalid

sosiouxme replied:
I agree it's a bit odd, however the alternatives seem about as odd (unless I'm missing some):

- hack this into the LimitRanger plugin, which would mean carrying a patch in Godeps
- run the LimitRanger before *and* after this plugin

Both are completely admin-configured, and seem like low risk for foot-shooting, although it certainly could confuse some.

liggitt commented on pkg/pod/admission/limit_request.go:
```diff
@@ -0,0 +1,98 @@
+func newPodLimitRequest(config io.Reader) (admission.Interface, error) {
+	parsed, err := ReadConfig(config)
+	return &podLimitRequest{
+		Handler:            admission.NewHandler(admission.Create, admission.Update),
```
what would this do on update? if it changes anything on the podspec, won't that fail to validate, since the spec is immutable?

sosiouxme replied:
good point, i guess there is no update on a pod that could affect these values.

liggitt commented on pkg/pod/admission/limit_request.go:
```diff
@@ -0,0 +1,98 @@
+	if !a.Enabled {
+		return nil
+	}
+	pod, ok := attr.GetObject().(*kapi.Pod)
```
are there subresources we need to ignore creates on?

sosiouxme replied:
confused. i was under the impression i'd just get a single admission request about the pod itself. Would there be multiple for the pod+subresources? I'm ignorant.

liggitt replied:
subresource API calls come through admission as well. in some cases (like posting to `pods/foo/status`, the API object send to the endpoint is actually a `*api.Pod` (so the type check above would succeed).

I'm pretty sure this admission controller should only be paying attention to creates on pods (not to updates, and not to subresources). You can check that by returning early if `len(attr.GetSubresource()) > 0`, among other ways

liggitt commented on pkg/pod/admission/limit_request.go:
```diff
@@ -0,0 +1,98 @@
+type limitRequestConfig struct {
+	Enabled                   bool    `json:"enabled"`
+	LimitCpuToMemoryRatio     float64 `json:"limitCpuToMemoryRatio"`
+	CpuRequestToLimitRatio    float64 `json:"cpuRequestToLimitRatio"`
```
s/Cpu/CPU/g

liggitt commented on pkg/pod/admission/limit_request.go:
```diff
@@ -0,0 +1,98 @@
+		resources := container.Resources
+		memLimit := resources.Limits.Memory()
+		if a.LimitCpuToMemoryRatio != 0.0 {
+			resources.Limits["cpu"] = resource.Quantity{
```
use the constants in k8s.io/kubernetes/pkg/api for "cpu" and "memory"

sosiouxme commented:
addressed comments, rebased, etc.

sosiouxme commented:
I still need to figure out how to get the adjusted numbers to come out with the right format...

sosiouxme commented:
@csrwng @deads2k @liggitt other than adding some integration testing I think this is done. Care to take another look?

spadgett commented:
@sosiouxme Looks like this is generating an invalid config.js.

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/12760011/33c70236-c9b2-11e5-8a3b-7116153613df.png)

sosiouxme commented:
@derekwaynecarr this is the override PR.

@liggitt derek notes that pod specs are mutable now and will be able to update at least pod limits and maybe requests too later - should I pre-emptively handle updates here too or wait until that functionality actually exists?

Also derek thinks this should run before the limitranger so that we don't create containers/pods that are adjusted after validation and would fail validation on the next update. I guess I agree, except that means this needs to fill in the limit defaults (if missing) before doing the overrides.

sosiouxme commented:
@spadgett sorry i'll fix that as soon as I can

spadgett commented:
@sosiouxme No worries, I have other limit validation stuff to fix :-/

liggitt commented on hack/test-cmd.sh:
```diff
@@ -142,7 +142,7 @@ os::cmd::expect_success_and_text "openshift ex validate master-config ${MASTER_C
 # breaking the config fails the validation check
 cp ${MASTER_CONFIG_DIR}/master-config.yaml ${BASETMPDIR}/master-config-broken.yaml
-os::util::sed '7,12d' ${BASETMPDIR}/master-config-broken.yaml
+os::util::sed '12,17d' ${BASETMPDIR}/master-config-broken.yaml
```
wouldn't a better way to break it be to append a duplicate key that points to a struct of the wrong type?
like `echo "
kubernetesMasterConfig: invalidData
" >> ${BASETMPDIR}/master-config-broken.yaml`

sosiouxme replied:
i think it's trying to test the validation with fields missing. yaml deserialization errors don't make the test pass. but it's not a bad idea to add a duplicate with fields missing rather than trying to hack up the existing file. i'll play with it.

liggitt commented on pkg/assets/handlers.go:
```diff
@@ -210,6 +212,13 @@ type WebConsoleConfig struct {
+	//   LimitCPUToMemoryRatio
+	//   CPURequestToLimitRatio
+	//   MemoryRequestToLimitRatio
+	LimitRequestOverrides api.PodLimitRequestConfig
```
only primitives here, this doesn't handle JSON serialization of sub objects

sosiouxme replied:
this doesn't get serialized, as such, rather put through a template.

csrwng replied:
I think it's causing the problem @spadgett pointed out above. You likely need to put the individual fields here.

sosiouxme replied:
no, that was me thinking template did more magic than it does :)

csrwng commented on pkg/cmd/server/kubernetes/master_config.go:
```diff
@@ -295,6 +286,22 @@ func BuildKubernetesMasterConfig(options configapi.MasterConfig, requestContextM
 	return kmaster, nil
 }
 
+// GetPluginConfigFile translates from the master plugin config to a file name containing
```
Please move to a common package like pkg/cmd/server/admission and call from https://github.com/openshift/origin/blob/master/pkg%2Fcmd%2Fserver%2Forigin%2Fmaster_config.go#L183 as well.

sosiouxme replied:
Moving it into pkg/cmd/server/admission/init.go - config is kinda init-related, right? Can be changed if that seems hokey.

csrwng commented on pkg/quota/admission/podlimitrequest/admission.go:
```diff
@@ -0,0 +1,125 @@
+	glog.V(5).Infof("PodLimitRequest about to read config:\n%v", configFile)
+	config := api.PodLimitRequestConfig{}
+	buffer := new(bytes.Buffer)
+	buffer.ReadFrom(configFile)
```
add check for error

sosiouxme replied:
oops

csrwng commented on pkg/quota/admission/podlimitrequest/admission.go:
```diff
@@ -0,0 +1,125 @@
+	buffer.ReadFrom(configFile)
+	glog.V(5).Infof("PodLimitRequest read config:\n%v", buffer.Bytes())
+	err := yaml.Unmarshal(buffer.Bytes(), &config)
+	if err == nil && config.Enabled {
```
if err != nil {
   return err
}
...

sosiouxme replied:
Well, it would have to return `config, error`. But it already does that. I read this as "if all is well and this plugin applies, do the logic. Otherwise return any errors." Does this really not seem more elegant, or is the "err != nil" convention really the only way to write this?

csrwng replied:
It's more idiomatic to return immediately instead of including the non-error case inside an if block
https://golang.org/doc/effective_go.html#if

csrwng commented on pkg/quota/admission/podlimitrequest/admission.go:
```diff
@@ -0,0 +1,125 @@
+	if !a.Config.Enabled {
+		return nil
+	}
+	if pod, isPod := attr.GetObject().(*kapi.Pod); isPod && attr.GetSubresource() == "" {
```
check for resource == kapi.Resource("pods") && getSubresource() == ""

it should also return immediately instead of containing the rest of the code inside the if block:

if ![cond] {
   return nil
}
blah...

sosiouxme replied:
> check for resource == kapi.Resource("pods") && getSubresource() == ""

is there a particular reason? because then, I'd have to do the cast separately.

> it should also return immediately instead of containing the rest of the code inside the if block:

if I'm not mistaken, @deads2k gave me the exact opposite recommendation https://github.com/openshift/origin/pull/6901#discussion-diff-51312510

csrwng replied:
I can think it'd be possible to have an API endpoint (say a virtual resource) that takes a pod object but doesn't mean it's creating a pod. I think checking the resource name instead of the object type narrows the scope of what you're accepting to "a pod getting created". @deads2k do you disagree?

sosiouxme replied:
i can see what you're saying here.

csrwng commented on pkg/quota/admission/podlimitrequest/admission.go:
```diff
@@ -0,0 +1,125 @@
+		return nil
+	}
+	if pod, isPod := attr.GetObject().(*kapi.Pod); isPod && attr.GetSubresource() == "" {
+		glog.V(8).Infof("PodLimitRequest is looking at creating a pod")
```
You may want to log the name/namespace of the pod getting created. Also use a loglevel <= 5.

sosiouxme replied:
ok good idea

csrwng commented on pkg/quota/admission/podlimitrequest/admission.go:
```diff
@@ -0,0 +1,125 @@
+		glog.V(8).Infof("PodLimitRequest is looking at creating a pod")
+
+		// allow annotations on project to override
+		if ns, err := a.KubeClient.Namespaces().Get(attr.GetNamespace()); err != nil {
```
you should likely use the project cache instead of retrieving the namespace directly here.

csrwng replied:
also return immediately after getting an error and remove the else below

sosiouxme replied:
you're right i don't want to do the lookup if i don't have to. i'm not familiar w/ using the project cache. do you happen to have a pointer to something that does that?

csrwng replied:
https://github.com/openshift/origin/blob/master/pkg/project/admission/requestlimit/admission.go 

Your plugin should implement the WantsProjectCache and Validator interfaces:
https://github.com/openshift/origin/blob/master/pkg/project/admission/requestlimit/admission.go#L60-L62
The code in master_config.go will call SetProjectCache on your plugin for you.

sosiouxme replied:
I thought those were only for openshift admission plugins...

csrwng replied:
It'll work for kube ones as well (at least the project cache)

sosiouxme replied:
I guess we could add the logic even when initing kube plugins? https://github.com/openshift/origin/blob/64082787228db4e85a33bc554b9e0506f5c8d707/pkg/cmd/server/kubernetes/master_config.go#L159 that seems complicated though

sosiouxme replied:
ah yes you pulled in the openshift initializer too https://github.com/openshift/origin/blob/64082787228db4e85a33bc554b9e0506f5c8d707/pkg/cmd/server/kubernetes/master_config.go#L166 cool!

csrwng commented on pkg/quota/admission/podlimitrequest/admission.go:
```diff
@@ -0,0 +1,125 @@
+)
+
+const (
+	annEnabled         = "openshift.io/PodLimitRequestEnabled"
```
s/annEnabled/podLimitRequestAnnotation/

csrwng commented on pkg/quota/admission/podlimitrequest/admission.go:
```diff
@@ -0,0 +1,125 @@
+		// allow annotations on project to override
+		if ns, err := a.KubeClient.Namespaces().Get(attr.GetNamespace()); err != nil {
+			glog.Warningf("PodLimitRequest got an error retrieving namespace: %v", err)
+			return err // this should not happen
```
errors from the plugin should generally be created as admission.NewForbidden(attributes, causingError)

sosiouxme replied:
ok good to know

sosiouxme commented:
@spadgett commit e130d70 should fix it.

spadgett commented:
@sosiouxme working now, thanks

derekwaynecarr commented on pkg/cmd/server/api/v1/types.go:
```diff
@@ -932,3 +935,19 @@ type AdmissionConfig struct {
+
+// PodLimitRequestConfig is the configuration for the PodLimitRequest
+// admission controller which overrides user-provided container request/limit values.
+type PodLimitRequestConfig struct {
```
Should this be PodContainerLimitRequestConfig instead?  I still have more to read here, but it won't be long until you can set requests/limits on the pod and what is in the container will become largely irrelevant...

sosiouxme replied:
Since this only enforces ratios, once that day comes could we just modify this controller to give pod req/limit the same treatment? Then the name would remain accurate and there would be no config changes. For density purposes we really only get to look at scheduling pods...

derekwaynecarr replied:
Ok.

derekwaynecarr commented on pkg/cmd/server/kubernetes/master_config.go:
```diff
@@ -36,12 +36,11 @@ import (
 
 // AdmissionPlugins is the full list of admission control plugins to enable in the order they must run
-var AdmissionPlugins = []string{"NamespaceLifecycle", "OriginPodNodeEnvironment", "LimitRanger", "ServiceAccount", "SecurityContextConstraint", "ResourceQuota", "SCCExecRestrictions"}
+var AdmissionPlugins = []string{"NamespaceLifecycle", "OriginPodNodeEnvironment", "LimitRanger", "PodLimitRequest", "ServiceAccount", "SecurityContextConstraint", "ResourceQuota", "SCCExecRestrictions"}
```
I don't like that this runs after `LimitRanger`, is there any guarantee that it will not exceed the allowed limits?

sosiouxme replied:
There isn't, my next task is to get it running before LimitRanger.

derekwaynecarr commented on pkg/quota/admission/podlimitrequest/admission.go:
```diff
@@ -0,0 +1,152 @@
+	}
+
+	// This relies on LimitRanger already having applied limits from the project.
+	// It will then adjust values as configured; not going to re-validate against LimitRanger.
```
not validating against LimitRanger will potentially cause pod updates to fail if your modifications violate the constraints on limit ranger.  Example pod updates:  change image name (not used), change active deadline seconds (used to cancel a build/deployer pod).

sosiouxme replied:
@derekwaynecarr going to run before LimitRanger so it validates. Confusing as it would be to an end user if their admin managed to mess this up, it seems like the right thing to do.

Implementation detail: I'm thinking I'll pick up any defaults, then run these ratio overrides, then pass off to LimitRanger. Of course the logic for merging in default limits is in LimitRanger (k8s.io/kubernetes/plugin/pkg/admission/limitranger/admission.go). It would be convenient for this if defaultContainerResourceRequirements and mergePodResourceRequirements were exported. Which approach seems best to you:

1. Capitalize these and use them. Carry the patch until we can get that change upstream.
2. Just use PodLimitFunc which already invokes these, and ignore any errors returned.
3. Copy those methods' source over to my plugin.

sosiouxme replied:
Picking 2 except will just run Admit() which is a one-liner

derekwaynecarr commented on pkg/quota/admission/podlimitrequest/admission.go:
```diff
@@ -0,0 +1,152 @@
+	// It will then adjust values as configured; not going to re-validate against LimitRanger.
+	for _, container := range pod.Spec.Containers {
+		resources := container.Resources
+		memLimit := resources.Limits.Memory()
```
what happens if this is undefined?

sosiouxme replied:
Limits.Memory() returns a pointer to a 0-value Quantity.

derekwaynecarr commented on pkg/quota/admission/podlimitrequest/admission_test.go:
```diff
@@ -0,0 +1,223 @@
+		name               string
+		config             api.PodLimitRequestConfig
+		object             runtime.Object
+		expectedMemRequest *inf.Dec
```
any reason you cant just use resource.Quantity here?  you can then use quantity.MustParse so its easier to read the test case then having to convert to the inf.Dec structure.

sosiouxme replied:
I'll give that a try. It would probably be clearer.

derekwaynecarr commented on pkg/quota/admission/podlimitrequest/admission_test.go:
```diff
@@ -0,0 +1,223 @@
+				{
+					Resources: kapi.ResourceRequirements{
+						Limits: kapi.ResourceList{
+							"cpu":    resource.Quantity{Amount: inf.NewDec(cpuLimit, 3)},
```
kapi.ResourceCPU

derekwaynecarr commented on pkg/quota/admission/podlimitrequest/admission_test.go:
```diff
@@ -0,0 +1,223 @@
+					Resources: kapi.ResourceRequirements{
+						Limits: kapi.ResourceList{
+							"cpu":    resource.Quantity{Amount: inf.NewDec(cpuLimit, 3)},
+							"memory": resource.Quantity{Amount: inf.NewDec(memLimit, 0)},
```
kap.ResourceMemory

derekwaynecarr commented on pkg/quota/admission/podlimitrequest/admission.go:
```diff
@@ -0,0 +1,152 @@
+	return nil
+}
+
+func (a *podLimitRequest) Admit(attr admission.Attributes) error {
```
Can you add a ```// TODO this will need to update when we have pod requests/limits``` so it's not lost?

sosiouxme commented on pkg/quota/admission/podlimitrequest/admission.go:
```diff
@@ -0,0 +1,160 @@
+
+	// Reuse LimitRanger logic to apply limit/req defaults from the project. Ignore validation
+	// errors, assume that LimitRanger will run after this plugin to validate.
+	a.LimitRanger.Admit(attr)
```
@derekwaynecarr so it turns out to be pretty easy to reuse LimitRanger like this and then have it also validate after. Look good?

derekwaynecarr replied:
looks good

sosiouxme commented:
[test]

smarterclayton commented on pkg/cmd/server/api/register.go:
```diff
@@ -58,9 +59,8 @@ func (*GoogleIdentityProvider) IsAnAPIObject()            {}
-func (*SessionSecrets) IsAnAPIObject() {}
-
-func (*LDAPSyncConfig) IsAnAPIObject() {}
+func (*PodLimitRequestConfig) IsAnAPIObject()             {}
```
These methods no longer have any meaning, all of them should be removed

smarterclayton commented on pkg/cmd/server/api/types.go:
```diff
@@ -980,3 +983,19 @@ type AdmissionConfig struct {
+type PodLimitRequestConfig struct {
+	unversioned.TypeMeta
+	// Enabled must be true for the plugin to do anything. It can be
+	// overridden per-project with the annotation openshift.io/PodLimitRequestEnabled
```
Annotation is not consistent with our pattern - "quota.openshift.io/pod-request-limit-enabled" is more accurate

sosiouxme replied:
worksforme

smarterclayton commented on pkg/cmd/server/start/master_args.go:
```diff
@@ -180,6 +183,19 @@ func (args MasterArgs) BuildSerializeableMasterConfig() (*configapi.MasterConfig
+	if podLimitRequest, err := admission.GetPluginConfigFile(kubernetesMasterConfig.AdmissionConfig.PluginConfig, "PodLimitRequest", ""); err != nil {
+		return nil, err
+	} else if podLimitRequest != "" {
+		configIo, err := os.Open(podLimitRequest)
```
What is a configIo?  This should be f or configFile

smarterclayton commented on pkg/cmd/server/api/types.go:
```diff
@@ -980,3 +983,19 @@ type AdmissionConfig struct {
+
+// PodLimitRequestConfig is the configuration for the PodLimitRequest
+// admission controller which overrides user-provided container request/limit values.
+type PodLimitRequestConfig struct {
```
This name doesn't describe what this is for - I'd prefer more useful information.  Something like PodOvercommitPolicy or PodRequestLimitOverride or ClusterOvercommitPolicy (which describes the intent of the config, not the mechanics).  If like the admission controller name to match the intent as well.  I'm open to other suggestions, but I don't think PodRequestLimit is enough.

sosiouxme replied:
I'm not enamored of the name either, just didn't have anything else come to mind. Something that expresses that the purpose is cluster-wide control of resource allocation, though this is just about pods at the moment. Frankly "RequestLimit"-anything probably doesn't lead the mind in the right direction. How about "ClusterResourceOverrides"?

smarterclayton replied:
ClusterResourceOverride works fine for the controller -
ClusterResourceOverridePolicy is probably appropriate for the config
*if* we will never have other resource policy in the same place.

sosiouxme replied:
I have no idea what future resource policy might look like, other than knowing that soon we'll be setting request/limit on the pod directly rather than containers. I could maybe see the scope of this plugin being expanded if other resources come into play? Not sure.

I went with `ClusterResourceOverride` and `ClusterResourceOverrideConfig`. Lemme know if you have any strong feelings about that.

smarterclayton replied:
That's fine

On Mon, Feb 8, 2016 at 3:03 PM, Luke Meyer <notifications@github.com> wrote:

> In pkg/cmd/server/api/types.go
> <https://github.com/openshift/origin/pull/6901#discussion_r52203403>:
>
> > @@ -980,3 +983,19 @@ type AdmissionConfig struct {
> >  	// on the master. Order is significant. If empty, a default list of plugins is used.
> >  	PluginOrderOverride []string
> >  }
> > +
> > +// PodLimitRequestConfig is the configuration for the PodLimitRequest
> > +// admission controller which overrides user-provided container request/limit values.
> > +type PodLimitRequestConfig struct {
>
> I have no idea what future resource policy might look like, other than
> knowing that soon we'll be setting request/limit on the pod directly rather
> than containers. I could maybe see the scope of this plugin being expanded
> if other resources come into play? Not sure.
>
> I went with ClusterResourceOverride and ClusterResourceOverrideConfig.
> Lemme know if you have any strong feelings about that.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6901/files#r52203403>.
>

smarterclayton commented on pkg/cmd/server/api/types.go:
```diff
@@ -980,3 +983,19 @@ type AdmissionConfig struct {
+	// overridden per-project with the annotation openshift.io/PodLimitRequestEnabled
+	Enabled bool
+	// LimitCPUToMemoryRatio (if > 0.0) pegs the CPU limit to a ratio of the memory limit;
+	// a base ratio of 1.0 scales CPU to 1000mcore per 1GiB of RAM.
```
Document here what happens to user provided values (and for the other types)

sosiouxme commented:
If anyone has further comments, let 'em rip, otherwise I'll try merging this later today or tomorrow.

deads2k commented on pkg/cmd/server/start/master_args.go:
```diff
@@ -182,6 +185,19 @@ func (args MasterArgs) BuildSerializeableMasterConfig() (*configapi.MasterConfig
 
 	dnsServingInfo := servingInfoForAddr(&dnsBindAddr)
 
+	overrideConfig := configapi.ClusterResourceOverrideConfig{}
```
I don't like this section.  The intent of this method is to build a serializeable config from arguments.  It should be entirely predictable and I don't think that it should allow an admission plugin config file to override the output of the actual fix section of the config at all.

@liggitt I've been out in rebase-land for a while.  Is this something you inspected already?

liggitt replied:
no, and I agree

sosiouxme replied:
The config file has to be made available to go into the web console asset bundle for the console to display appropriate creation options; this was part of the feature specification. That bundle is built from the AssetConfig in the MasterConfig, and this is where that's defined. I don't think there's any other reason for the AssetConfig to even be in there; I would agree there should probably be some other way to transmit this info but at the moment there isn't. I wasn't going for rewriting anything I don't have to.

the admission controller itself has no actual access to the master config when initialized; it has only the admission plugin config and a filename. there is no way to insert things into the asset configurati

sosiouxme replied:
Let me connect the dots now that I have a little more time to pay attention.

The task is to get this plugin's config out to the console. That console config is provided over here https://github.com/sosiouxme/origin/blob/31c763839b1dbf4c7e6612582a9131d43952f3e0/pkg/cmd/server/origin/asset.go#L200 where it really only has the AssetConfig object to work with.

Said AssetConfig object is basically just a container for configapi.AssetConfig https://github.com/sosiouxme/origin/blob/31c763839b1dbf4c7e6612582a9131d43952f3e0/pkg/cmd/server/origin/asset_config.go which it is given by start_master.StartAPI https://github.com/sosiouxme/origin/blob/31c763839b1dbf4c7e6612582a9131d43952f3e0/pkg/cmd/server/start/start_master.go#L454

So alright, I guess I can just expand the origin.AssetConfig object https://github.com/sosiouxme/origin/blob/31c763839b1dbf4c7e6612582a9131d43952f3e0/pkg/cmd/server/origin/asset_config.go#L8 with a plugin config member and have StartAPI fill that in rather than putting it in the serializable MasterConfig.AssetConfig object. Would that be preferable?

deads2k commented on pkg/cmd/server/api/types.go:
```diff
@@ -998,3 +1001,26 @@ type AdmissionConfig struct {
+	// The plugin's actions can be disabled per-project with the project annotation
+	// quota.openshift.io/cluster-resource-override-enabled="false", so cluster admins
+	// can exempt infrastructure projects and such from the overrides.
+	Enabled bool
```
`Enabled` seems wrong.  If you've got the config, you're enabled.  If you don't, then you're not.  Having this suggests that what you really want is a pointer to this type in `AssetConfig`

liggitt replied:
I agree... @csrwng was advocating a style where admission plugins are in the chain, but dormant based on config flags... I don't think I like that

csrwng replied:
I think that having a way to enable/disable via config is the right thing to do. Whether we do that via a flag or by having the config present/absent, I have no preference. For plugins like this one where the ordering is critical, I'd rather have in the default chain instead of asking users to insert it in the right place in their configs.

sosiouxme replied:
Enabled seemed simplest because the null value is false; this carries over nicely to sending the same object to the web console.

You also might simply want to turn a plugin off temporarily without completely wiping out its config. I think it's a useful pattern.

deads2k replied:
> I agree... @csrwng was advocating a style where admission plugins are in the chain, but dormant based on config flags in their respective config files... I don't think I like that

I like them in the chain, but rather than an "enable" flag, I'd go based on presence or absence.

deads2k replied:
> You also might simply want to turn a plugin off temporarily without completely wiping out its config. I think it's a useful pattern.

If you want that power, use version control on your config or reference a separate file.

deads2k replied:
> Enabled seemed simplest because the null value is false; this carries over nicely to sending the same object to the web console.

The web console tolerates its absence.

sosiouxme replied:
I'll go with config presence/absence then.

liggitt replied:
@csrwng, are there other plugins already using the `enabled: true` style? If so, I'd like to have the same debate for them, and be consistent.

csrwng replied:
Only one. The RunOnceDuration plugin. I can submit a PR to use absence of config to mean that it's disabled.

csrwng replied:
Thinking about it some more, @deads2k @liggitt I don't like the absence of config to mean that the plugin is disabled. That would mean that an empty config vs a nil config mean totally different things. My preference would be that an empty config mean the same thing as a nil config.

deads2k replied:
> nking about it some more, @deads2k @liggitt I don't like the absence of config to mean that the plugin is disabled. That would mean that an empty config vs a nil config mean totally different things. My preference would be that an empty config mean the same thing as a nil config.

I'm ok with that too. An empty config is detectable.

csrwng replied:
How should we detect it in this case without the Enabled flag? just 0's for all ratios?

deads2k replied:
> How should we detect it in this case without the Enabled flag? just 0's for all ratios?

It would an empty file when the admission plugin gets the filename, right?

csrwng replied:
That's not what I meant to ask :-) but I think I answered my own question. It looks like this plugin does nothing if the config is empty because all ratios are 0, making the Enabled flag superfluous.

smarterclayton replied:
I'd like a stronger signal than all zeros.  That's fragile the first
time someone adds a new field.

csrwng replied:
So... we keep the Enabled flag?

deads2k replied:
> So... we keep the Enabled flag?

I see
 1. no entry in the plugin config - disable
 1. no value in location and config - fails validation
 1. no value in location, value in config - this is enabled.
 1. value in location, value in config - fails validation
 1 . value in location, no value in config - this is enabled

Which case is ambiguous?  "value in location, no value in config, zero bytes in file"?  That should fail.

csrwng replied:
So are you saying that we will never have an empty config, and therefore it's ok to say only if config == nil then we disable the plugin?

How about a file (or inline config) with
```
apiVersion: v1
kind: ClusterResourceOverrideConfig
```

deads2k replied:
> How about a file (or inline config) with

Enabled and you get what you deserve.

sosiouxme replied:
Yeah in that case actually this admission controller will fail validation at initialization, because while you might deserve it, we can easily point out that you shot your foot.

I think I can get on board with no config == nil pointer. Specifically including a member Enabled always felt a bit redundant, and in my case was intended mainly to prevent having to check if all members are 0 (in two places). Nil pointer (or in JS, missing entry) does the job just as well.

deads2k commented on hack/test-cmd.sh:
```diff
@@ -142,7 +142,7 @@ os::cmd::expect_success_and_text "openshift ex validate master-config ${MASTER_C
 # breaking the config fails the validation check
 cp ${MASTER_CONFIG_DIR}/master-config.yaml ${BASETMPDIR}/master-config-broken.yaml
-os::util::sed '7,12d' ${BASETMPDIR}/master-config-broken.yaml
+echo "kubernetesMasterConfig: {}" >> ${BASETMPDIR}/master-config-broken.yaml
```
Previously, there was an `apiVersion` and `kind` that allowed for proper decoding.  I would expect this change to break the decoding path and fail earlier than intended.  I think that this change actually skips the bit where we're validating the config because I don't see how it could construct the object correctly.

sosiouxme replied:
If you add this line at the end, the config is loaded and fails validation like so:

    start_master.go:271] kubernetesMasterConfig.proxyClientInfo: Invalid value: "": if no client certificate is specified, TLS pods and services cannot validate requests came from the proxy

Is that sufficient or could you suggest something more correct to fail validation as intended?

deads2k replied:
> Is that sufficient or could you suggest something more correct to fail validation as intended?

That's sufficient and unexpected.  It probably means I've hardcoded the destination type.  hrm...

deads2k commented on pkg/cmd/server/admission/init.go:
```diff
@@ -38,3 +40,19 @@ func Validate(plugins []admission.Interface) error {
+// GetPluginConfigFile translates from the master plugin config to a file name containing
+// a particular plugin's config (the file may be a temp file if config is embedded)
+func GetPluginConfigFile(pluginConfig map[string]api.AdmissionPluginConfig,
+	pluginName string, defConfigFile string) (string, error) {
```
Is `defConfigFile` the content of the file or its location?  (adjust the name to indicate what it is).

Also, that seems a little early for a line break, 80 characters is a relic.

sosiouxme replied:
perhaps there should be a coding style document specifying at what point a line break is indicated? I was just keeping that one line from sticking out from the rest of the file...

deads2k commented on pkg/cmd/server/api/types.go:
```diff
@@ -998,3 +1001,26 @@ type AdmissionConfig struct {
+
+// ClusterResourceOverrideConfig is the configuration for the ClusterResourceOverride
+// admission controller which overrides user-provided container request/limit values.
+type ClusterResourceOverrideConfig struct {
```
This type is specifically related to quota configuration right?  Is there some intent to generalize it?  If not, this name should reflect its intent.

sosiouxme replied:
Not really quota-related per se. It's for ops to tweak overcommit mostly.

Clayton and I had a brief discussion at https://github.com/openshift/origin/pull/6901#discussion-diff-52099169 - I would be happy for something that's more descriptive. Maybe one of your epic half-page names is in order :)

deads2k replied:
> I would be happy for something that's more descriptive. Maybe one of your epic half-page names is in order :)

Well, there's a reason 80 characters is too short. :)

This seems to be focused specifically on limits.  I think I would have it match your variable name `LimitRequestOverridesConfig`.

sosiouxme replied:
On Tue, Feb 9, 2016 at 8:04 AM, David Eads <notifications@github.com> wrote:
>
> This seems to be focused specifically on limits. I think I would have it
> match your variable name LimitRequestOverridesConfig.
>

@smarterclayton or anyone else want to chime in? I don't have a strong
opinion - hard to think of any name that won't be at least mildly
misleading to somebody. But I only want to change this one more time this
week :)

deads2k replied:
> @smarterclayton or anyone else want to chime in? I don't have a strong
> opinion - hard to think of any name that won't be at least mildly
> misleading to somebody. But I only want to change this one more time this
> week :)

I request something closer.  A `ClusterResourceOverrideConfig` that doesn't mutate a `Cluster`-scoped thing or even general `Resource` is pretty hard to defend.

sosiouxme replied:
> I request something closer. A ClusterResourceOverrideConfig that doesn't
> mutate a Cluster-scoped thing or even general Resource is pretty hard to
> defend.
>
>
LimitRequestOverrides{,Config} going once... going twice...

smarterclayton replied:
You are mutating resources across the entire cluster.  I would prefer
you leave the name as is.

deads2k replied:
> You are mutating resources across the entire cluster.  I would prefer
> you leave the name as is.

So does every other admission plugin we have (`NamespaceLifecycle`, `OriginPodNodeEnvironment`, `LimitRanger`, `ServiceAccount`, ***).  None of them are selectively choosing namespaces, so they all operate on resources across the cluster.  `Cluster` in this sense is a zero entropy phrase that adds no value to understanding.

deads2k commented on pkg/cmd/server/admission/init.go:
```diff
@@ -38,3 +40,19 @@ func Validate(plugins []admission.Interface) error {
+
+// GetPluginConfigFile translates from the master plugin config to a file name containing
+// a particular plugin's config (the file may be a temp file if config is embedded)
+func GetPluginConfigFile(pluginConfig map[string]api.AdmissionPluginConfig,
```
This doesn't look like the correct package for this function.  Why not put it with the function you're delegating to?

sosiouxme replied:
sounds like a good idea

deads2k commented on pkg/cmd/server/api/types.go:
```diff
@@ -998,3 +1001,26 @@ type AdmissionConfig struct {
+	// LimitCPUToMemoryRatio (if > 0.0) overrides the CPU limit to a ratio of the memory limit;
+	// a base ratio of 1.0 scales CPU to 1000mcore per 1GiB of RAM. This is done before
+	// overriding the CPU request.
+	LimitCPUToMemoryRatio float64
```
This should be a human readable file.  I would have expected percentages.  Has that already been considered and rejected?

sosiouxme replied:
No, actually it didn't occur to me :) But you're right, ops would probably prefer that.

deads2k commented on pkg/cmd/server/kubernetes/master_config.go:
```diff
@@ -147,16 +144,9 @@ func BuildKubernetesMasterConfig(options configapi.MasterConfig, requestContextM
-				if err != nil {
-					return nil, err
-				}
+			configFile, err := oadmission.GetPluginConfigFile(options.KubernetesMasterConfig.AdmissionConfig.PluginConfig, pluginName, server.AdmissionControlConfigFile)
```
I prefer the old factorization, but if @csrwng prefers this I can live with it.

sosiouxme replied:
> I prefer the old factorization, but if @csrwng prefers this I can live with it.

Since the web console asset config also needs to read this file, I thought it would be best for both to share the function to do that. Not sure if you're advocating for duplicating this code instead, or just pulling the func back into this file or something.

deads2k commented:
tldr; Not ready for merge.

If I'm interpreting correctly, the intent is to provide some optional values to the web console based on whether and how a particular admission controller that lives on kubernetes is configured.  If that's the case, I'm pretty sure this is wired in a way that allows them to get out of sync.  Also, if that's the intent, you should wire them together in that order (not the reverse) and make the linkage obvious.

You should also note and build a test that demonstrates, how this attempt at configuration will not be correctly tied together if we ever get an external kube.  With that in mind, have you made an issue upstream to figure out how to resolve the problem of dependencies on the admission chain?

deads2k commented on pkg/quota/admission/clusterresourceoverride/admission.go:
```diff
@@ -0,0 +1,177 @@
+	return config, err
+}
+
+func Validate(config api.ClusterResourceOverrideConfig) error {
```
Move this method to a `validation` package under your `api` package.

deads2k commented:
In terms of the API objects, I think this should be left out of `configapi.AssetConfig` because that forces an admin to repeat himself.  Instead, I would add the values as a pointer to the `origin.AssetConfig`.  That `struct` is intended to be the non-serializeable config and can hold state.  You can initialize the value based on the admission plugin value for the `KubernetesMasterConfig`.  That eliminates the "repeat yourself" problem, correctly indicates the dependency, and gives us the flexibility to change it later since its not permanently serialized on disk.

I could see us revisiting this choice if the dual knowledge becomes necessary at a later date.

deads2k commented:
@liggitt you probably have an opinion on: https://github.com/openshift/origin/pull/6901#issuecomment-181856975

deads2k commented:
Promoting into the main thread so we don't lose it:

> Let me connect the dots now that I have a little more time to pay attention.
> 
> The task is to get this plugin's config out to the console. That console config is provided over here https://github.com/sosiouxme/origin/blob/31c763839b1dbf4c7e6612582a9131d43952f3e0/pkg/cmd/server/origin/asset.go#L200 where it really only has the AssetConfig object to work with.
> 
> Said AssetConfig object is basically just a container for configapi.AssetConfig https://github.com/sosiouxme/origin/blob/31c763839b1dbf4c7e6612582a9131d43952f3e0/pkg/cmd/server/origin/asset_config.go which it is given by start_master.StartAPI https://github.com/sosiouxme/origin/blob/31c763839b1dbf4c7e6612582a9131d43952f3e0/pkg/cmd/server/start/start_master.go#L454
> 
> So alright, I guess I can just expand the origin.AssetConfig object https://github.com/sosiouxme/origin/blob/31c763839b1dbf4c7e6612582a9131d43952f3e0/pkg/cmd/server/origin/asset_config.go#L8 with a plugin config member and have StartAPI fill that in rather than putting it in the serializable MasterConfig.AssetConfig object. Would that be preferable?

I'd give Jordan time to read/comment, but that would be my preference.  The type can still be precise, but it would be looked up from the admission plugin part of the config to avoid repetition.  The alternative would require cross config validation.

sosiouxme commented:
Ha! You were right there ahead of me. OK, I'll go ahead with that, no problem.

sosiouxme commented on pkg/cmd/server/start/start_master.go:
```diff
@@ -451,7 +451,7 @@ func StartAPI(oc *origin.MasterConfig, kc *kubernetes.MasterConfig) error {
 	var standaloneAssetConfig *origin.AssetConfig
 	if oc.WebConsoleEnabled() {
-		config, err := origin.BuildAssetConfig(*oc.Options.AssetConfig)
+		config, err := origin.NewAssetConfig(*oc.Options.AssetConfig, oc.Options.KubernetesMasterConfig.AdmissionConfig.PluginConfig)
```
@deads2k @liggitt OK this is easy to back out if you don't like but I thought it would be good to pass in the config for all admission plugins rather than scope it to one. Briefly considered passing in the entire master config, decided on this scope for now.

deads2k replied:
How about passing through the exact config you're interested in rather than the entire map?  That would keep the knowledge of how these things are specified segregated.

This is the sort of interaction I was expecting.

sosiouxme replied:
I thought it not unlikely that perhaps another admission controller might also have config that goes into the web console asset bundle. But *shrug* it is not too hard to just make that change when is needed. I can pass just the config for the one plugin we care about now.

sosiouxme commented:
> You should also note and build a test that demonstrates, how this attempt at configuration will not be correctly tied together if we ever get an external kube. With that in mind, have you made an issue upstream to figure out how to resolve the problem of dependencies on the admission chain?

@deads2k I am not sure I follow. If we had an external kubernetes cluster, I imagine OpenShift would still run all of the same admission controllers that it does with embedded kube, and then pass the result to kube which would run its own set of admission controllers and actually create the object. I could be wrong about how that would work, though, not being that familiar with the guts of running the embedded kube. Wouldn't any kube admission controller that's only present in OpenShift code be untenable then?

Or maybe this is irrelevant with the config object now in the proper place.

liggitt commented:
> If we had an external kubernetes cluster, I imagine OpenShift would still run all of the same admission controllers that it does with embedded kube, and then pass the result to kube which would run its own set of admission controllers and actually create the object.

Admission is wired into the individual rest handlers, not as a single filter like authn/authz. Individual rest handlers do not exist with external kube, only a blanket proxy of a given sub path, so origin admission plugins never see kube resources when running against external kube.

sosiouxme commented:
On Tue, Feb 9, 2016 at 8:27 PM, Jordan Liggitt <notifications@github.com>
wrote:
>
> Admission is wired into the individual rest handlers, not as a single
> filter like authn/authz. Individual rest handlers do not exist with
> external kube, only a blanket proxy of a given sub path, so origin
> admission plugins never see kube resources when running against external
> kube.
>
>
That does sound like any admission controllers we add in origin for kube
resources would be inert against an external kube; so the question of
configuration seems moot. These plugins would have to be upstreamed if we
wanted them working in that scenario (which is probably not even possible,
viz project cache), or we'd have to hack something (presumably ugly) into
the proxy logic to put kube resources through origin admission first.

sosiouxme commented:
@deads2k I think I've addressed all of your recommendations, aside from the name change (https://github.com/openshift/origin/pull/6901#discussion_r52551606). Maybe you and @smarterclayton could hash that out in the office and let me know :)

deads2k commented on pkg/assets/handlers.go:
```diff
@@ -187,6 +188,13 @@ window.OPENSHIFT_CONFIG = {
   	oauth_client_id: "{{ .OAuthClientID | js}}",
   	logout_uri: "{{ .LogoutURI | js}}"
   },
+  {{ with .LimitRequestOverrides }}
```
Can you make a test to confirm this works nicely with `nil`?

sosiouxme replied:
On Thu, Feb 11, 2016 at 9:22 AM, David Eads <notifications@github.com>
wrote:

> In pkg/assets/handlers.go
> <https://github.com/openshift/origin/pull/6901#discussion_r52606444>:
>
> > +  {{ with .LimitRequestOverrides }}
>
> Can you make a test to confirm this works nicely with nil?
>

done with
https://github.com/sosiouxme/origin/commit/3896a9fc858c68fab4af6ec8252f42ee34c989d4

deads2k commented on pkg/cmd/server/start/start_master.go:
```diff
@@ -451,7 +454,20 @@ func StartAPI(oc *origin.MasterConfig, kc *kubernetes.MasterConfig) error {
 	if oc.WebConsoleEnabled() {
-		config, err := origin.BuildAssetConfig(*oc.Options.AssetConfig)
+		var overrideConfig *overrideapi.ClusterResourceOverrideConfig = nil
+		if overridePluginConfigFile, err := pluginconfig.GetPluginConfigFile(oc.Options.KubernetesMasterConfig.AdmissionConfig.PluginConfig, "ClusterResourceOverride", ""); err != nil {
```
Yours is the first needing this kind of a ref.  Let's go with a constant for the magic string to ensure we don't slip.

deads2k commented on pkg/quota/admission/clusterresourceoverride/admission.go:
```diff
@@ -0,0 +1,161 @@
+// newClusterResourceOverride returns an admission controller for containers that
+// configurably overrides container resource request/limits
+func newClusterResourceOverride(client kclient.Interface, config io.Reader) (admission.Interface, error) {
+	glog.V(5).Infof("ClusterResourceOverride admission controller is loaded")
```
This is extraneous, you'll always hit the one below.

deads2k commented on pkg/quota/admission/clusterresourceoverride/admission.go:
```diff
@@ -0,0 +1,161 @@
+	glog.V(5).Infof("ClusterResourceOverride admission controller is loaded")
+	parsed, err := ReadConfig(config)
+	glog.V(5).Infof("ClusterResourceOverride admission controller got config: %v\nerror: (%T) %[2]v", parsed, err)
+	return &clusterResourceOverridePlugin{
```
If you get an error, I'd rather not return a value that can be used, no matter how improperly.  Add 
```go
if err != nil{
    return nil, err
}
```
above this and then end with an explicit `, nil`.

deads2k commented on pkg/quota/admission/clusterresourceoverride/admission.go:
```diff
@@ -0,0 +1,161 @@
+	a.ProjectCache = projectCache
+}
+
+func ReadConfig(configFile io.Reader) (*api.ClusterResourceOverrideConfig, error) {
```
This looks a lot like `configapilatest.ReadYaml` only not quite right.  Can you replace with that function?

sosiouxme replied:
yes, done, thanks to your help

deads2k commented on pkg/quota/admission/clusterresourceoverride/admission.go:
```diff
@@ -0,0 +1,161 @@
+		return nil, err
+	}
+	config := &api.ClusterResourceOverrideConfig{}
+	err := yaml.Unmarshal(buffer.Bytes(), config)
```
I think this whole function is actually `ReadYaml`, but this is doing the wrong thing.  You're unmarshalling versioned data into an unversioned object without conversions.  This is getting lucky because the types match and the default json serializer doesn't care about case.  It's also avoiding proper type identification, so it won't fail even when it should.

sosiouxme replied:
On Thu, Feb 11, 2016 at 9:33 AM, David Eads <notifications@github.com>
wrote:

> In pkg/quota/admission/clusterresourceoverride/admission.go
> <https://github.com/openshift/origin/pull/6901#discussion_r52607805>:
>
> I think this whole function is actually ReadYaml, but this is doing the
> wrong thing. You're unmarshalling versioned data into an unversioned object
> without conversions. This is getting lucky because the types match and the
> default json serializer doesn't care about case. It's also avoiding proper
> type identification, so it won't fail even when it should.
>

Yeah, I didn't know much about reading config when I started this and
didn't have a good example, so I just winged it and it seemed to work.

Replacing with `ReadYaml` my tests start failing getting a 0-val object.
This may be because I've screwed up something about type registration; that
whole thing is still a bit of a mystery to me. I may need some pointers to
resolve this.

deads2k commented on pkg/quota/admission/clusterresourceoverride/admission.go:
```diff
@@ -0,0 +1,161 @@
+
+// TODO this will need to update when we have pod requests/limits
+func (a *clusterResourceOverridePlugin) Admit(attr admission.Attributes) error {
+	glog.V(8).Infof("ClusterResourceOverride admission controller is invoked")
```
Try avoid going below 6.  The client debugging gets really noisy past 6 and we made it that high to avoid it spamming detailed logs from other components.

sosiouxme replied:
Did you mean "try to avoid going above 6"?

deads2k replied:
> Did you mean "try to avoid going above 6"?

Use 1-6.

deads2k commented on pkg/quota/admission/clusterresourceoverride/admission.go:
```diff
@@ -0,0 +1,161 @@
+	// Reuse LimitRanger logic to apply limit/req defaults from the project. Ignore validation
+	// errors, assume that LimitRanger will run after this plugin to validate.
+	glog.V(5).Infof("ClusterResourceOverride: initial pod limits are: %#v", pod.Spec.Containers[0].Resources)
+	if err := a.LimitRanger.Admit(attr); err != nil {
```
@derekwaynecarr does double defaulting make anything weird happen?  I see this as another reason to split defaulting from authorizing.

sosiouxme replied:
shouldn't, but i agree it would be cleaner to split the functions out.

derekwaynecarr replied:
@deads2k I agree that this is a reason for splitting defaulting from rejecting in the future.

derekwaynecarr replied:
As you review this PR, can you add to this issue what you would like to see changed in LimitRange plug-in to support embedded use scenario.  https://github.com/kubernetes/kubernetes/issues/21077

deads2k commented on pkg/quota/admission/clusterresourceoverride/admission.go:
```diff
@@ -0,0 +1,161 @@
+	if err := a.LimitRanger.Admit(attr); err != nil {
+		glog.V(5).Infof("ClusterResourceOverride: error from LimitRanger: %#v", err)
+	}
+	pod, ok = attr.GetObject().(*kapi.Pod)
```
Is this pod different from the previous pod you had?  I would have expected your manipulation to actually adjust the values being pointed to so that this was unnecessary.

sosiouxme replied:
errr, no, i added that in when i was confused about why limitranger didn't seem to be actually setting defaults (which was just my own confusion). thanks for finding that.

deads2k commented on pkg/quota/admission/clusterresourceoverride/admission.go:
```diff
@@ -0,0 +1,161 @@
+		// for CPU there is some hocus-pocus with the "scale" component when multiplying
+		// (inf.Dec does not have a multiplication method).
+		if a.Config.LimitCPUToMemoryPercent != 0.0 {
+			resources.Limits[kapi.ResourceCPU] = resource.Quantity{
```
So it looks to me like bad things will happen if my project has no `ResourceQuota` associated with it to actually provide the defaults.  Shouldn't these only be applied if you actually have the limits?

sosiouxme replied:
On Thu, Feb 11, 2016 at 9:47 AM, David Eads <notifications@github.com>
wrote:
>
> So it looks to me like bad things will happen if my project has no
> ResourceQuota associated with it to actually provide the defaults.
> Shouldn't these only be applied if you actually have the limits?
>
> The Memory() method returns a zero-value Quantity if missing. Same for
Cpu(). Then limit/request would be rewritten based on that.

@derekwaynecarr do you happen to know if a zero quantity is treated as "no
limit" or "limit 0 i.e. you get nothing"? I see now I was assuming the
former without verifying.

derekwaynecarr replied:
To be clear, `ResourceQuota` does not do defaulting.

This has problems if the other things that do defaulting did not give value (i.e. `InitialResources` or `LimitRange`).  

Either this plug-in can do some defaulting itself, or it should reject if it sees a pod that gave no memory limit.  I think I prefer rejection.

sosiouxme replied:
On Thu, Feb 11, 2016 at 1:22 PM, Derek Carr <notifications@github.com>
wrote:

> This has problems if the other things that do defaulting did not give
> value (i.e. InitialResources or LimitRange).
>

For the case that a project has no LimitRange defaults,  I changed the
logic not to override based on a missing value.

Either this plug-in can do some defaulting itself, or it should reject if
> it sees a pod that gave no memory limit. I think I prefer rejection.
>

At the moment, at least, if limits are specified in a LimitRange, then so
are defaults. So it seems to me the only scenario where this would happen
is when there aren't any limits set; that might be admin error but it seems
strange to me to reject for missing pod limits when there aren't any limits
in a LimitRange to be violated.

deads2k commented on pkg/quota/admission/clusterresourceoverride/admission.go:
```diff
@@ -0,0 +1,161 @@
+	return &clusterResourceOverridePlugin{
+		Handler:     admission.NewHandler(admission.Create),
+		Config:      parsed,
+		LimitRanger: limitranger.NewLimitRanger(client, wrapLimit),
```
This is adding another list/watch and cache instead of re-using the existing one.  In addition, it looks like this admission plugin logically requires the other one to activated.  Is there a reasonable way to wire these together?  @derekwaynecarr  is the too expensive to allow?  I'm inclined to allow a followup issue unless you think this is really problematic.

I see two followups:
 1. Some additional `Wants` interface that can say that you must have some sort of admission plugin.  The list should be split for before and after in the chain.
 1. Share a limit ranger or at least the cache to avoid coherence problems.

sosiouxme replied:
This is adding another list/watch and cache instead of re-using the
> existing one. In addition, it looks like this admission plugin logically
> requires the other one to activated. Is there a reasonable way to wire
> these together?
>
Not that I can see. You could combine the two plugins together. Or you
could split LimitRanger into two plugins to go before and after in the
chain. Both would require upstream work that would probably not happen in a
timely manner, or carrying a patch.

The only other thing that comes to mind is adding a Wants interface to
share the limits cache as with the project cache.

derekwaynecarr replied:
@deads2k @sosiouxme - we are really starting to hit the limits on what admission control was originally designed to provide, and what we are now pushing it to support.  there are a number of issues of architectural purity that get complicated when we want to support embedded kube (where we run a separate set of admission controllers, but want to share caches potentially) versus when we run separate from kube (where we do not want to share caches).  A future TODO where we can figure out how to support a `Wants/Provides` type of semantic around shared caches in admission controllers is probably a good thing to explore when we look to optimize the embedded kube model more, but I doubt this is the straw that breaks the camels back. 

Our co-located controllers definitely have duplicate caches today as well, so its a general problem.

deads2k commented on pkg/quota/admission/clusterresourceoverride/api/register.go:
```diff
@@ -0,0 +1,56 @@
+// availableVersions lists all known external versions for this group from most preferred to least preferred
+var availableVersions = []unversioned.GroupVersion{v1.SchemeGroupVersion}
+
+func init() {
```
This should be in an `install` package

sosiouxme replied:
>
> This should be in an install package
>

OK. Is there a reason or is this just convention?

deads2k replied:
> OK. Is there a reason or is this just convention?

Reason.  If you agree that types should be opt-in enabled (this may be questionable, but its the upstream assertion), then you need a way to depend on the types without enabling them at the same time.  This would allow you to wire this into the dependency chain of something like master start, but have it gated.   As you currently have it structured, that power has been removed.

deads2k commented on pkg/quota/admission/clusterresourceoverride/api/validate/validate.go:
```diff
@@ -0,0 +1,27 @@
+func Validate(config *api.ClusterResourceOverrideConfig) error {
+	if config != nil {
+		if config.LimitCPUToMemoryPercent == 0.0 && config.CPURequestToLimitPercent == 0.0 && config.MemoryRequestToLimitPercent == 0.0 {
+			return fmt.Errorf("%s plugin enabled but no ratios specified", pluginname)
```
Use the normal `field.Invalid` sort of errors like other validation packages.

deads2k commented on pkg/quota/admission/clusterresourceoverride/api/validate/validate.go:
```diff
@@ -0,0 +1,27 @@
+package validate
```
`validation` by convention.

derekwaynecarr commented on pkg/quota/admission/clusterresourceoverride/admission.go:
```diff
@@ -0,0 +1,161 @@
+	glog.V(5).Infof("ClusterResourceOverride: pod limits after LimitRanger are: %#v", pod.Spec.Containers[0].Resources)
+	for _, container := range pod.Spec.Containers {
+		resources := container.Resources
+		memLimit := resources.Limits.Memory()
```
you should use form:

```
memLimit, found := resources.Limits[api.ResourceMemory]
```

This would let you distinguish if the value was set or if you are just getting the nil 0 value.

derekwaynecarr commented on pkg/quota/admission/clusterresourceoverride/admission.go:
```diff
@@ -0,0 +1,161 @@
+		// with 3 places of accuracy for millicores. In order to maintain 3 places of accuracy
+		// for CPU there is some hocus-pocus with the "scale" component when multiplying
+		// (inf.Dec does not have a multiplication method).
+		if a.Config.LimitCPUToMemoryPercent != 0.0 {
```
this needs to not be a float.  floats are evil.

derekwaynecarr commented on pkg/quota/admission/clusterresourceoverride/admission.go:
```diff
@@ -0,0 +1,161 @@
+		// (inf.Dec does not have a multiplication method).
+		if a.Config.LimitCPUToMemoryPercent != 0.0 {
+			resources.Limits[kapi.ResourceCPU] = resource.Quantity{
+				Amount: inf.NewDec(int64(float64(memLimit.Value())*cpuBaseScaleFactor*a.Config.LimitCPUToMemoryPercent/100.0), 3),
```
I think this may be prone to losing precision since value loses fractional parts.  

Ideally, quantity would expose `Mul` from the underlying `inf.Dec` internal value, but I would defer to that instead of what is happening here...

sosiouxme replied:
I missed that Mul even exists. Maybe because I wasn't looking to multiply by an `inf.Dec`

sosiouxme replied:
I don't think there was any fractional part to lose from memory (fraction of a byte?). But I'm switching to use `Mul` anyway because it requires less hand-waving to explain.

derekwaynecarr commented on pkg/quota/admission/clusterresourceoverride/admission.go:
```diff
@@ -0,0 +1,161 @@
+		if a.Config.LimitCPUToMemoryPercent != 0.0 {
+			resources.Limits[kapi.ResourceCPU] = resource.Quantity{
+				Amount: inf.NewDec(int64(float64(memLimit.Value())*cpuBaseScaleFactor*a.Config.LimitCPUToMemoryPercent/100.0), 3),
+				Format: resources.Limits.Cpu().Format,
```
Do not depend on that value right now, the upstream stuff in resource helpers makes no format guarantee:
https://github.com/kubernetes/kubernetes/blob/master/pkg/api/resource_helpers.go#L29

I will open an upstream PR so it guarantees to default to the right format, but specify `DecimalSI` directly here.

derekwaynecarr replied:
https://github.com/kubernetes/kubernetes/pull/21088

derekwaynecarr commented on pkg/quota/admission/clusterresourceoverride/admission.go:
```diff
@@ -0,0 +1,161 @@
+		if a.Config.MemoryRequestToLimitPercent != 0.0 {
+			resources.Requests[kapi.ResourceMemory] = resource.Quantity{
+				Amount: inf.NewDec(int64(float64(resources.Limits.Memory().Value())*a.Config.MemoryRequestToLimitPercent/100.0), 0),
+				Format: resources.Requests.Memory().Format,
```
see comment above, specify `BinarySI` directly.

derekwaynecarr commented on pkg/quota/admission/clusterresourceoverride/api/types.go:
```diff
@@ -0,0 +1,22 @@
+	//
+	// LimitCPUToMemoryPercent (if > 0) overrides the CPU limit to a ratio of the memory limit;
+	// 100% overrides CPU to 1 core per 1GiB of RAM. This is done before overriding the CPU request.
+	LimitCPUToMemoryPercent float64
```
we need to make these integers i think , and its up to use to come up with a precision range that we want if 1-100 is insufficient.

sosiouxme replied:
What if I made it an `inf.Dec`... then I could actually use Mul...

sosiouxme commented on pkg/cmd/server/kubernetes/master_config.go:
```diff
@@ -43,7 +43,7 @@ import (
 
 // AdmissionPlugins is the full list of admission control plugins to enable in the order they must run
-var AdmissionPlugins = []string{"NamespaceLifecycle", "OriginPodNodeEnvironment", "LimitRanger", "ServiceAccount", "SecurityContextConstraint", "BuildDefaults", "BuildOverrides", "ResourceQuota", "SCCExecRestrictions"}
+var AdmissionPlugins = []string{"NamespaceLifecycle", "OriginPodNodeEnvironment", "ClusterResourceOverride", "LimitRanger", "ServiceAccount", "SecurityContextConstraint", "BuildDefaults", "BuildOverrides", "ResourceQuota", "SCCExecRestrictions"}
```
@deads2k should this be the sole constant in the list?

deads2k replied:
> @deads2k should this be the sole constant in the list?

For now, yes.  They should all be updated, but yours has more referrers than the rest so I'd like it to change now.  I don't think you should have to change them all in this pull, just make an issues and we'll followup.

sosiouxme commented on pkg/quota/admission/clusterresourceoverride/admission.go:
```diff
@@ -0,0 +1,176 @@
+		if memFound && a.Config.LimitCPUToMemoryPercent != 0 {
+			resources.Limits[kapi.ResourceCPU] = resource.Quantity{
+				// float math is necessary here as there is no way to create an inf.Dec to represent cpuBaseScaleFactor < 0.001
+				Amount: multiply(inf.NewDec(int64(float64(memLimit.Value())*cpuBaseScaleFactor), 3), a.internalConfig.limitCPUToMemoryRatio),
```
@derekwaynecarr just wanted to run this by you in case you see a better way

derekwaynecarr replied:
I don't see a better way.  If you have a good test that covers large memory values for what is realistically possible, then I think we are fine.

sosiouxme replied:
>
> I don't see a better way.  If you have a good test that covers large memory values for what is realistically possible, then I think we are fine.
>
>
I added some tests for large and small values. There is some loss of
precision when you get down below millicores but I think that's unavoidable
with this library, and not significant to users.

sosiouxme commented:
@deads2k @derekwaynecarr I think I've addressed all your comments so far.

@dobbymoodge suggested that it would probably be a good idea to keep the overridden values from dropping below the min limits, as this would unnecessarily frustrate users with rejections when e.g. their project has a min memory limit of 128Mi, they specify a container memory limit that's 128Mi, and the ratio is 50% overriding request to 64Mi which then is rejected. Since this plugin already has a limits cache in order to init `LimitRanger` I might as well use it to floor overridden limits at min to prevent this kind of situation. Let me know if you see any issues with that.

I'll be squashing/rebasing and working on that last thing while I wait for any more concerns to be expressed.

sosiouxme commented:
Shoot, I don't have the limits cache, it gets created in LimitRanger and can't be passed in or exported :( So if I wanted to implement the limit floor I'd have to make yet another of the same cache...

liggitt commented:
probably ripe for pulling out to be shared

sosiouxme commented:
@liggitt indeed https://github.com/kubernetes/kubernetes/issues/21077 but I didn't get the impression @derekwaynecarr was going to get that done and merged real soon.

I could patch an export function into LimitRanger but bleah.  I wonder if this concern is enough to warrant building another cache with attendant watch, or if I should just let it slide until we have a shareable one. Probably let it slide.

sosiouxme commented:
[test]

sosiouxme commented on assets/app/scripts/app.js:
```diff
@@ -279,7 +279,7 @@ angular
   .constant("LOGGING_URL", (window.OPENSHIFT_CONFIG || {}).loggingURL)
   .constant("METRICS_URL", (window.OPENSHIFT_CONFIG || {}).metricsURL)
-  .constant("LIMIT_REQUEST_OVERRIDES", _.get(window.OPENSHIFT_CONFIG, "limitRequestOverrides", {}))
+  .constant("LIMIT_REQUEST_OVERRIDES", _.get(window.OPENSHIFT_CONFIG, "limitRequestOverrides", null))
```
@spadgett I think I did this right but I can't figure out where to view the changes on my console.

sosiouxme replied:
nm, found it and all seems to be working. so just an FYI for you.

spadgett replied:
Just get rid of the 3rd arg to `_.get`, and it will be undefined. Or you could leave it `{}` and remove the null checks in limits.js.

sosiouxme commented:
Anyone have feelings about a merge? Devcut being tomorrow and all...

sosiouxme commented:
same flake https://github.com/openshift/origin/issues/6031 on both from https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/1170/ ... hmm.

[test]

deads2k commented on pkg/cmd/server/api/install/install.go:
```diff
@@ -12,6 +12,7 @@ import (
 	_ "github.com/openshift/origin/pkg/build/admission/defaults/api/install"
 	_ "github.com/openshift/origin/pkg/build/admission/overrides/api/install"
 	_ "github.com/openshift/origin/pkg/project/admission/requestlimit/api/install"
+	_ "github.com/openshift/origin/pkg/quota/admission/clusterresourceoverride/api/install"
```
these were moved to `pkg/cmd/server/api/install/install.go`

sosiouxme replied:
On Tue, Feb 16, 2016 at 9:15 AM, David Eads <notifications@github.com>
wrote:

> In pkg/cmd/server/api/install/install.go
> <https://github.com/openshift/origin/pull/6901#discussion_r53014883>:
>
> > +	_ "github.com/openshift/origin/pkg/quota/admission/clusterresourceoverride/api/install"
>
> these were moved to pkg/cmd/server/api/install/install.go
>

I followed suit and put mine in the same place as the other admission
plugins... did I miss something?

deads2k replied:
> I followed suit and put mine in the same place as the other admission
> plugins... did I miss something?

Nope, seems I lost my mind.

deads2k commented on pkg/quota/admission/clusterresourceoverride/admission.go:
```diff
@@ -0,0 +1,176 @@
+}
+type clusterResourceOverridePlugin struct {
+	*admission.Handler
+	internalConfig
```
avoid anonymous inclusion.  It creates unexpected side effects

deads2k replied:
This looks like a structure that's parallel to `Config`.  Why do I have both?  Why not just `internalConfig`?

sosiouxme replied:
On Tue, Feb 16, 2016 at 9:18 AM, David Eads <notifications@github.com>
wrote:
>
> > +type clusterResourceOverridePlugin struct {
> > +	*admission.Handler
> > +	internalConfig
>
> This looks like a structure that's parallel to Config. Why do I have
> both? Why not just internalConfig?
>
>
Good point, it's redundant now. Thanks.

deads2k commented on pkg/quota/admission/clusterresourceoverride/admission.go:
```diff
@@ -0,0 +1,176 @@
+	}
+	glog.V(5).Infof("%s admission controller loaded with config: %v", api.PluginName, parsed)
+	internal := internalConfig{
+		limitCPUToMemoryRatio:     inf.NewDec(parsed.LimitCPUToMemoryPercent, 2),
```
I don't know what this does, but I'm assuming it was one of @derekwaynecarr's suggestions.

sosiouxme replied:
On Tue, Feb 16, 2016 at 9:17 AM, David Eads <notifications@github.com>
wrote:
>
> > +		limitCPUToMemoryRatio:     inf.NewDec(parsed.LimitCPUToMemoryPercent, 2),
>
> I don't know what this does, but I'm assuming it was one of
> @derekwaynecarr <https://github.com/derekwaynecarr>'s suggestions.
>
>

He didn't want me using floats and fiddling directly with inf.Dec scale
factors. So this converts an integer percent to an inf.Dec ratio (scales
down by 2 places) so I can actually use inf.Mul to multiply.

deads2k commented on pkg/quota/admission/clusterresourceoverride/admission.go:
```diff
@@ -0,0 +1,176 @@
+// newClusterResourceOverride returns an admission controller for containers that
+// configurably overrides container resource request/limits
+func newClusterResourceOverride(client kclient.Interface, config io.Reader) (admission.Interface, error) {
+	parsed, err := ReadConfig(config)
```
I'm not seeing a validation check.  Did I just miss it?

deads2k replied:
> I'm not seeing a validation check. Did I just miss it?

Ah, I see.  Why should we allow the silent construction of an invalid admission plugin instance?

sosiouxme replied:
On Tue, Feb 16, 2016 at 9:23 AM, David Eads <notifications@github.com>
wrote:
>
> Ah, I see. Why should we allow the silent construction of an invalid
> admission plugin instance?
>

The plugin init manager runs validation right after and kicks it out if it
fails. Doing validation here seemed redundant, but I can if you like.

sosiouxme replied:
On Tue, Feb 16, 2016 at 9:29 AM, Luke Meyer <lmeyer@redhat.com> wrote:
>
> The plugin init manager runs validation right after and kicks it out if it
> fails. Doing validation here seemed redundant, but I can if you like.
>

Actually since the original redundant Config has been removed from the
plugin object, doing validation here became essential.

deads2k commented:
Relatively minor comments.  Someone else will need to review the asset stuff, probably @liggitt

liggitt commented:
@spadgett  for asset review

spadgett commented on assets/app/scripts/services/limits.js:
```diff
@@ -34,30 +34,30 @@ angular.module("openshiftConsole")
-      return LIMIT_REQUEST_OVERRIDES.enabled &&
-             LIMIT_REQUEST_OVERRIDES.limitCPUToMemoryRatio &&
-             LIMIT_REQUEST_OVERRIDES.cpuRequestToLimitRatio;
+      return LIMIT_REQUEST_OVERRIDES !== null &&
```
```
      return LIMIT_REQUEST_OVERRIDES &&
             LIMIT_REQUEST_OVERRIDES.limitCPUToMemoryPercent &&
             LIMIT_REQUEST_OVERRIDES.cpuRequestToLimitPercent;
```

spadgett commented on assets/app/scripts/services/limits.js:
```diff
@@ -34,30 +34,30 @@ angular.module("openshiftConsole")
 
     var isRequestCalculated = function(computeResource) {
-      if (!LIMIT_REQUEST_OVERRIDES.enabled) {
+      if (LIMIT_REQUEST_OVERRIDES === null) {
```
You can to catch `undefined`, too, not just `null`.

```
        if (!LIMIT_REQUEST_OVERRIDES) {
          return false;
        }
```

spadgett commented on assets/app/scripts/services/limits.js:
```diff
@@ -34,30 +34,30 @@ angular.module("openshiftConsole")
+      return LIMIT_REQUEST_OVERRIDES !== null &&
              computeResource === 'cpu' &&
-             LIMIT_REQUEST_OVERRIDES.limitCPUToMemoryRatio;
+             LIMIT_REQUEST_OVERRIDES.limitCPUToMemoryPercent;
```
```
        return LIMIT_REQUEST_OVERRIDES &&
               computeResource === 'cpu' &&
               LIMIT_REQUEST_OVERRIDES.limitCPUToMemoryPercent;
```

sosiouxme commented:
@deads2k can I get a LGTM? Tests actually passed finally. Will of course squash/rebase.

deads2k commented:
lgtm as long as @spadgett is happy with the asset stuff.

spadgett commented:
Asset changes LGTM

sosiouxme commented:
[merge]
